addons/gamification/models/gamification_challenge.py

```diff
@@ -61,233 +62,218 @@ class gamification_challenge(osv.Model):
-            help="Never reminded if no value or zero is specified."),
-        'last_report_date': fields.date('Last Report Date'),
-        'next_report_date': fields.function(_get_next_report_date,
-            type='date', string='Next Report Date', store=True),
-
-        'category': fields.selection(lambda s, *a, **k: s._get_categories(*a, **k),
-            string="Appears in", help="Define the visibility of the challenge through menus", required=True),
-        }
-
-    _defaults = {
-        'period': 'once',
-        'state': 'draft',
-        'visibility_mode': 'personal',
-        'report_message_frequency': 'never',
-        'last_report_date': fields.date.today,
-        'manager_id': lambda s, cr, uid, c: uid,
-        'category': 'hr',
-        'reward_failure': False,
-        'report_template_id': lambda s, *a, **k: s._get_report_template(*a, **k),
-        'reward_realtime': True,
-    }
-
-
-    def create(self, cr, uid, vals, context=None):
-        """Overwrite the create method to add the user of groups"""
+                challenge.next_report_date = False
 
+    def _update_vals(self, vals):
         if vals.get('user_domain'):
-            user_ids = self._get_challenger_users(cr, uid, vals.get('user_domain'), context=context)
-
+            user_ids = self._get_challenger_users(vals.get('user_domain'))
             if not vals.get('user_ids'):
                 vals['user_ids'] = []
             vals['user_ids'] += [(4, user_id) for user_id in user_ids]
+        return vals
 
-        return super(gamification_challenge, self).create(cr, uid, vals, context=context)
-
-    def write(self, cr, uid, ids, vals, context=None):
-        if isinstance(ids, (int,long)):
-            ids = [ids]
-
-        if vals.get('user_domain'):
-            user_ids = self._get_challenger_users(cr, uid, vals.get('user_domain'), context=context)
-
-            if not vals.get('user_ids'):
-                vals['user_ids'] = []
-            vals['user_ids'] += [(4, user_id) for user_id in user_ids]
+    @api.model
+    def create(self, vals):
+        """Overwrite the create method to add the user of groups"""
+        return super(GamificationChallenge, self).create(self._update_vals(vals))
 
-        write_res = super(gamification_challenge, self).write(cr, uid, ids, vals, context=context)
+    @api.multi
+    def write(self, vals):
+        write_res = super(GamificationChallenge, self).write(self._update_vals(vals))
 
         if vals.get('report_message_frequency', 'never') != 'never':
             # _recompute_challenge_users do not set users for challenges with no reports, subscribing them now
-            for challenge in self.browse(cr, uid, ids, context=context):
-                self.message_subscribe(cr, uid, [challenge.id], [user.partner_id.id for user in challenge.user_ids], context=context)
+            for challenge in self:
+                challenge.message_subscribe([user.partner_id.id for user in challenge.user_ids])
```

Comment:
here you can make use of mapped function this way : 
 [user.partner_id.id for user in challenge.user_ids] ==> challenge.mapped('user_ids.partner_id').ids